File: scripts/export-document/ir-to-pptx.py
#!/usr/bin/env python3
"""Document IR JSON → editable PPTX (python-pptx)."""
from __future__ import annotations

import argparse
import json
import os
from pptx import Presentation
from pptx.util import Inches, Pt

def normalize_inline(text: str) -> str:
    value = str(text or "")
    value = value.replace("**", "").replace("__", "").replace("`", "")
    import re
    value = re.sub(r"\[([^\]]+)\]\([^)]+\)", r"\1", value)
    value = re.sub(r"\[([^\]]+)\]\[[^\]]*\]", r"\1", value)
    value = re.sub(r"\s+", " ", value).strip()
    return value


def add_slide_title_body(prs: Presentation, title: str, lines: list[str]) -> None:
    layout = prs.slide_layouts[1] if len(prs.slide_layouts) > 1 else prs.slide_layouts[0]
    slide = prs.slides.add_slide(layout)
    if slide.shapes.title:
        slide.shapes.title.text = normalize_inline(title)
    body = slide.placeholders[1] if len(slide.placeholders) > 1 else None
    if body is not None:
        tf = body.text_frame
        tf.clear()
        for i, line in enumerate(lines):
            p = tf.paragraphs[0] if i == 0 else tf.add_paragraph()
            p.text = normalize_inline(line)
            p.font.size = Pt(18)


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--ir", required=True)
    parser.add_argument("--output", required=True)
    args = parser.parse_args()

    with open(args.ir, encoding="utf-8") as f:
        ir = json.load(f)

    prs = Presentation()
    slide_lines: list[str] = []
    slide_title = ir.get("title") or "Presentation"

    def flush_slide() -> None:
        nonlocal slide_lines, slide_title
        if slide_lines:
            add_slide_title_body(prs, slide_title, slide_lines)
            slide_lines = []

    for block in ir.get("blocks") or []:
        btype = block.get("type")
        if btype == "slideBreak":
            flush_slide()
            continue
        if btype == "heading":
            if slide_lines:
                flush_slide()
            slide_title = block.get("text") or slide_title
            continue
        if btype == "paragraph":
            slide_lines.append(block.get("text") or "")
        elif btype == "blockquote":
            slide_lines.append(block.get("text") or "")
        elif btype == "list":
            for item in block.get("items") or []:
                prefix = f"{item}" if block.get("ordered") else f"• {item}"
                slide_lines.append(prefix)
        elif btype == "table":
            flush_slide()
            blank = prs.slide_layouts[6]
            slide = prs.slides.add_slide(blank)
            headers = block.get("headers") or []
            rows = block.get("rows") or []
            text = " | ".join(headers) + "\n" + "\n".join(" | ".join(r) for r in rows)
            box = slide.shapes.add_textbox(Inches(0.5), Inches(0.8), Inches(12), Inches(5))
            box.text_frame.text = text
        elif btype == "image" and block.get("resolvedPath") and os.path.isfile(block["resolvedPath"]):
            flush_slide()
            blank = prs.slide_layouts[6]
            slide = prs.slides.add_slide(blank)
            slide.shapes.add_picture(block["resolvedPath"], Inches(0.5), Inches(0.5), width=Inches(12))
        elif btype == "chartImage" and block.get("resolvedPath") and os.path.isfile(block["resolvedPath"]):
            flush_slide()
            blank = prs.slide_layouts[6]
            slide = prs.slides.add_slide(blank)
            slide.shapes.add_picture(block["resolvedPath"], Inches(0.5), Inches(0.5), width=Inches(12))

    flush_slide()
    if len(prs.slides) == 0:
        add_slide_title_body(prs, slide_title, ["(empty)"])

    os.makedirs(os.path.dirname(os.path.abspath(args.output)) or ".", exist_ok=True)
    prs.save(args.output)
    print(args.output)


if __name__ == "__main__":
    main()

File: scripts/export-document/ir-to-pptx.mjs
#!/usr/bin/env node
/**
 * Document IR JSON → editable PPTX (pptxgenjs fallback when python-pptx unavailable).
 */
import { readFileSync, mkdirSync } from 'node:fs';
import path from 'node:path';
import pptxgen from 'pptxgenjs';

const inlineModule = await import(
  path.resolve(process.cwd(), 'dist/src/saas/document-export/markdownInline.js')
).catch(() => null);
const normalizeMarkdownInline = inlineModule?.normalizeMarkdownInline ?? ((text) => String(text ?? ''));

const args = process.argv.slice(2);
function readArg(flag) {
  const i = args.indexOf(flag);
  return i >= 0 ? args[i + 1] : undefined;
}

const irPath = readArg('--ir');
const output = readArg('--output');
if (!irPath || !output) {
  console.error('Usage: ir-to-pptx.mjs --ir <ir.json> --output <file.pptx>');
  process.exit(1);
}

const ir = JSON.parse(readFileSync(path.resolve(irPath), 'utf8'));
const pres = new pptxgen();
pres.layout = 'LAYOUT_WIDE';

let slideTitle = ir.title || 'Presentation';
let bodyLines = [];

function flushSlide() {
  if (bodyLines.length === 0) return;
  const slide = pres.addSlide();
  slide.addText(normalizeMarkdownInline(slideTitle), {
    x: 0.5,
    y: 0.35,
    w: 12.3,
    h: 0.9,
    fontSize: 28,
    bold: true,
    color: '363636',
  });
  slide.addText(bodyLines.map((line) => normalizeMarkdownInline(line)).join('\n'), {
    x: 0.5,
    y: 1.35,
    w: 12.3,
    h: 5.8,
    fontSize: 18,
    color: '363636',
    valign: 'top',
  });
  bodyLines = [];
}

for (const block of ir.blocks || []) {
  switch (block.type) {
    case 'slideBreak':
      flushSlide();
      break;
    case 'heading':
      flushSlide();
      slideTitle = block.text || slideTitle;
      break;
    case 'paragraph':
      bodyLines.push(block.text || '');
      break;
    case 'blockquote':
      bodyLines.push(block.text || '');
      break;
    case 'list':
      for (const item of block.items || []) {
        bodyLines.push(block.ordered ? `${item}` : `• ${item}`);
      }
      break;
    case 'table': {
      flushSlide();
      const slide = pres.addSlide();
      const headers = block.headers || [];
      const rows = block.rows || [];
      const text = normalizeMarkdownInline(
        (headers.length ? headers.join(' | ') + '\n' : '') +
        rows.map((row) => row.join(' | ')).join('\n'),
      );
      slide.addText(text, {
        x: 0.5,
        y: 0.8,
        w: 12.3,
        h: 6,
        fontSize: 14,
        color: '363636',
        valign: 'top',
      });
      break;
    }
    case 'image':
    case 'chartImage': {
      const imagePath = block.resolvedPath;
      if (!imagePath) break;
      flushSlide();
      const slide = pres.addSlide();
      slide.addImage({
        path: imagePath,
        x: 0.5,
        y: 0.5,
        w: 12.3,
        h: 6.5,
        sizing: { type: 'contain', w: 12.3, h: 6.5 },
      });
      break;
    }
    default:
      break;
  }
}

flushSlide();
if (pres.slides.length === 0) {
  const slide = pres.addSlide();
  slide.addText(normalizeMarkdownInline(slideTitle), { x: 0.5, y: 0.35, w: 12.3, h: 0.9, fontSize: 28, bold: true });
  slide.addText('(empty)', { x: 0.5, y: 1.35, w: 12.3, h: 5.8, fontSize: 18 });
}

const outPath = path.resolve(output);
mkdirSync(path.dirname(outPath), { recursive: true });
await pres.writeFile({ fileName: outPath });
console.log(outPath);
